routes/slots.js

```diff
@@ -1,34 +1,214 @@
+              slot: r.name,
+              conflictGroup: conflictGroup.name
+            });
+            count = count + 1;
+            if (count === conflictData.length) {
+              res.status(200).json({
+                passDataId: passDataId,
+                conflictDataName: conflictDataName,
+                groupOption: groupOption
+              });
+            }
+          });
+        });
+      }else {
+        res.status(200).json({
+          passDataId: passDataId,
+          conflictDataName: conflictDataName,
+          groupOption: groupOption
+        });
+      }
+    });
+  });
+});
+
+/*
+ return json data:
+ {
+   passDataId: // slot Ids can be removed
+   conflictDataName: {
+     slot: // conflict slot name
+   }
+ }
+ */
+slots.post('/removeGroupValidate',auth.ensureAuthenticated, function (req, res) {
+  var passDataId = [];
+  var conflictDataName = [];
+  var count = 0;
+  // validate slot
+  Slot.find({
+    '_id': {$in: req.body.slotIds}
+  }, function (err, docs) {
+    if (err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    docs.forEach(function (d) {
+      if (d.inGroup) {
+        passDataId.push(d._id);
+      } else {
+        conflictDataName.push({
+          slot: d.name
+        });
+      }
+      count = count + 1;
+      if (count === docs.length) {
+        res.status(200).json({
+          passDataId: passDataId,
+          conflictDataName: conflictDataName
+        });
+      }
+    });
+  });
+});
+
+slots.post('/addGroup',auth.ensureAuthenticated, function (req, res) {
+  // change slots of slotGroup
+  SlotGroup.findOne({'name': req.body.slotGroupName}, function(err, slotGroup){
+    if(err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    slotGroup.slots = slotGroup.slots.concat(req.body.slotIds);// concat array
+    slotGroup.save(function(err) {
+      if(err) {
+        console.error(err);
+        return res.status(500).send(err.message);
+      }
+      // change inGroup of slot
+      Slot.update({ '_id': {$in: req.body.slotIds}}, {inGroup: slotGroup._id}, {multi: true}, function(err) {
+        if(err) {
+          console.error(err);
+          return res.status(500).send(err.message);
+        }
+        res.status(200).end();
+      })
+    });
+  });
+});
+
+slots.post('/removeGroup',auth.ensureAuthenticated, function (req, res) {
+  // delete items in inGroup field of slot
+  Slot.update({ '_id': {$in: req.body.slotIds}}, {inGroup: null}, {multi: true}, function(err) {
+    if(err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    // delete specified slots of slotGroup
+    SlotGroup.find(function(err, docs){
+      if(err) {
+        console.error(err);
+        return res.status(500).send(err.message);
+      }
+      var count = 0;
+      docs.forEach(function(slotGroup) {
+        slotGroup.slots = slotGroup.slots.filter(function (x) {
```

Comment:
use pull to remove from the set see https://docs.mongodb.com/manual/reference/operator/update/pull/